Engine/DataFeeds/ZipDataCacheProvider.cs

```diff
@@ -170,26 +160,80 @@ public void Dispose()
         {
-            var clearCacheIfOlderThan = DateTime.UtcNow.AddSeconds(-CacheSeconds);
-
-            // clean all items that that are older than CacheSeconds than the current date
-            foreach (var zip in _zipFileCache)
+            // we just want one thread cleaning the cache at the same time
+            if (Monitor.TryEnter(_zipFileCache))
             {
-                if (zip.Value.Uncache(clearCacheIfOlderThan))
+                try
                 {
-                    // removing it from the cache
-                    CachedZipFile removed;
-                    if (_zipFileCache.TryRemove(zip.Key, out removed))
+                    var clearCacheIfOlderThan = utcNow.AddSeconds(-CacheSeconds);
+                    // clean all items that that are older than CacheSeconds than the current date
+                    foreach (var zip in _zipFileCache)
                     {
-                        // disposing zip archive
-                        removed.Dispose();
+                        if (zip.Value.Uncache(clearCacheIfOlderThan))
+                        {
+                            // only clear items if they are not being used
+                            if (Monitor.TryEnter(zip.Value))
+                            {
+                                try
+                                {
+                                    // removing it from the cache
+                                    CachedZipFile removed;
+                                    if (_zipFileCache.TryRemove(zip.Key, out removed))
+                                    {
+                                        // disposing zip archive
+                                        removed.Dispose();
+                                    }
+                                }
+                                finally
+                                {
+                                    Monitor.Exit(zip.Value);
+                                }
+                            }
+                        }
                     }
                 }
+                finally
+                {
+                    Monitor.Exit(_zipFileCache);
+                }
+
+                _nextCacheScan = utcNow.AddSeconds(CacheSeconds);
             }
+        }
+
+        private Stream CacheAndCreateStream(string filename, string entryName, DateTime utcNow)
+        {
+            Stream stream = null;
+            var dataStream = _dataProvider.Fetch(filename);
 
-            _lastCacheScan = DateTime.UtcNow;
+            if (dataStream != null)
+            {
+                try
+                {
+                    var newItem = new CachedZipFile(dataStream, filename, utcNow);
+
+                    // here we don't need to lock over the cache item
+                    // because it was still not added in the cache
+                    stream = CreateStream(newItem, entryName, filename);
+
+                    if (!_zipFileCache.TryAdd(filename, newItem))
```

Comment:
`ConcurrentDictionary<TKey, TValue>` provides an `AddOrUpdate` method that may be better suited here. Another pattern I've seen in the past is to use `ConcurrentDictionary<TKey, Lazy<TValue>>` which provides thread-safety on object creation by virtue of the `Lazy<T>`. What we have here _will_ work, but we'll also be creating and disposing an entry randomly when we don't need to. Using the `Lazy<T>` will prevent this case from happening.

When using the `ConcurrentDictionary<TKey, Lazy<TValue>>` you would use
```csharp
_zipFileCache.GetOrAdd(filename, fname => new Lazy<Stream>(() => /** lazy stream creation**/)
```

and we'll be guaranteed that the returned `Lazy<Stream>` is the same one used by all thread. We'll also be guaranteed that only a _single_ stream will be created. [Here's](https://blogs.endjin.com/2015/10/using-lazy-and-concurrentdictionary-to-ensure-a-thread-safe-run-once-lazy-loaded-collection/) an article I found describing the pattern. Also of interest, is the [Remarks](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.getoradd?view=netframework-4.8) section of the `ConcurrentDictionary` MSDN entry for `GetOrAdd` describing the locking methodology used which guarantees that the returned instance is the same for all thread _but does not_ guarantee that only a single delegate is executed, hence the usage of `Lazy<T>` to ensure we only execute the factory delegate once.